lib/_http_outgoing.js

```diff
@@ -832,6 +833,10 @@ OutgoingMessage.prototype.end = function end(chunk, encoding, callback) {
     }
     return this;
   } else if (!this._header) {
+    if (this.socket) {
```

Comment:
Also instead of duplicating the logic block to catch both cases:
```js
if (this.socket) {
 this.socket.cork();
}
```
maybe it would be better to add a single block on top before the cases:
```js
if (this.socket && !this.finished && (chunk || !this._header)) { // not finished, socket exists and data will be written
  this.socket.cork();
}
```

Reply:
Yea, I think this whole function needs some refactoring. I'd prefer to do that as a separate PR.

Reply:
Fine by me!